src/proj.rs

```diff
@@ -0,0 +1,791 @@
+            0.0, 0.0, 0.0,-xf2,-yf2,-1.0, y2*xf2, y2*yf2, y2,
+            xf2, yf2, 1.0, 0.0, 0.0, 0.0,-x2*xf2,-x2*yf2,-x2,
+            0.0, 0.0, 0.0,-xf3,-yf3,-1.0, y3*xf3, y3*yf3, y3,
+            xf3, yf3, 1.0, 0.0, 0.0, 0.0,-x3*xf3,-x3*yf3,-x3,
+            0.0, 0.0, 0.0,-xf4,-yf4,-1.0, y4*xf4, y4*yf4, y4,
+            xf4, yf4, 1.0, 0.0, 0.0, 0.0,-x4*xf4,-x4*yf4,-x4,
+            xf4, yf4, 1.0, 0.0, 0.0, 0.0,-x4*xf4,-x4*yf4,-x4]);
+                
+        match a.svd() {
+            Err(_) => None,
+            Ok((s, _, v)) => {
+                if s[[8,8]].abs() > 0.0001 || // rank(a) must be 8
+                   s[[7,7]].abs() < 0.0001 {  // but not 7
+                    None
+                } else {
+                    let h = v.col(8).into_matrix().into_vec();
+                    let mut transform = 
+                        [h[0] as f32, h[1] as f32, h[2] as f32,
+                         h[3] as f32, h[4] as f32, h[5] as f32,
+                         h[6] as f32, h[7] as f32, h[8] as f32];
+                    transform = normalize(transform);
+                    let class = TransformationClass::Projection;
+
+                    try_inverse(&transform)
+                        .map(|inverse| { Proj { transform, inverse, class }})
+                }
+            }
+        }
+    }
+
+    // Helper functions used as optiomization in warp
+
+    #[inline(always)]
+    fn warp_p(&self, x: f32, y: f32) -> (f32, f32) {
+        let t = &self.transform;
+        let d = t[6]*x + t[7]*y + t[8];
+
+        ((t[0]*x + t[1]*y + t[2])/d,
+         (t[3]*x + t[4]*y + t[5])/d)
+    }
+
+    #[inline(always)]
+    fn warp_a(&self, x: f32, y: f32) -> (f32, f32) {
+        let t = &self.transform;
+
+        ((t[0]*x + t[1]*y + t[2]),
+         (t[3]*x + t[4]*y + t[5]))
+    }
+
+    #[inline(always)]
+    fn warp_t(&self, x: f32, y: f32) -> (f32, f32) {
+        let t = &self.transform;
+        let tx = t[2];
+        let ty = t[5];
+
+        (x + tx,
+         y + ty)
+    }
+}
+
+/// Combines two projective transformations
+/// (matrix multiplication)
+impl Mul<Proj> for Proj {
+    type Output = Proj;
+
+    fn mul(self, rhs: Proj) -> Proj {
+        use TransformationClass as TC;
+        let t = mul3x3(self.transform, rhs.transform);
+        let i = mul3x3(rhs.inverse, self.inverse);
+
+        let class = match (self.class, rhs.class) {
+            (TC::Translation, TC::Translation) => TC::Translation,
+            (TC::Translation, TC::Affine) => TC::Affine,
+            (TC::Affine, TC::Translation) => TC::Affine,
+            (TC::Affine, TC::Affine)      => TC::Affine,
+            (_, _) => TC::Projection
+        };
+
+        Proj { transform: t, inverse: i, class: class }
+    }
+}
+
+/// Combines two projective transformations
+/// (matrix multiplication)
+impl<'a, 'b> Mul<&'b Proj> for &'a Proj {
+    type Output = Proj;
+
+    fn mul(self, rhs: &Proj) -> Proj {
+        use TransformationClass as TC;
```

Comment:
You can define one of these two `Mul` implementations in terms of the other, and remove the duplication.